File: test-suite/js/security/redos-regex-buggy.ts
type RequestLike = {
  query: Record<string, string | undefined>;
  body: Record<string, string | undefined>;
  nextUrl: { searchParams: URLSearchParams };
};

export function queryPattern(req: RequestLike): RegExp {
  const pattern = req.query.pattern ?? "";
  return new RegExp(pattern, "i");
}

export function nextSearchPattern(request: RequestLike): RegExp {
  const raw = request.nextUrl.searchParams.get("filter") ?? "";
  return RegExp(raw);
}

export function destructuredBodyPattern(req: RequestLike): RegExp {
  const { matcher } = req.body;
  return new RegExp(matcher ?? ".*");
}

export function templateInterpolatedPattern(req: RequestLike): RegExp {
  const pattern = req.query.prefix ?? "";
  return new RegExp(`^${pattern}:[a-z0-9_-]+$`);
}

File: test-suite/golang/security/redos_regex_buggy.go
package security

import (
	"net/http"
	"regexp"

	"github.com/go-chi/chi/v5"
)

func queryRegex(r *http.Request) (*regexp.Regexp, error) {
	pattern := r.URL.Query().Get("pattern")
	return regexp.Compile(pattern)
}

func routeRegex(r *http.Request) *regexp.Regexp {
	slug := chi.URLParam(r, "slug")
	return regexp.MustCompile(slug)
}

func headerMatch(req *http.Request, value string) bool {
	pattern := req.Header.Get("X-Filter")
	matched, _ := regexp.MatchString(pattern, value)
	return matched
}
